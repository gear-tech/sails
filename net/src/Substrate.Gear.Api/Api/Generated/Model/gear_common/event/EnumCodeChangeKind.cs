#nullable disable
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using Substrate.NetApi.Model.Types.Base;
using System.Collections.Generic;


namespace Substrate.Gear.Api.Generated.Model.gear_common.@event
{
    
    
    /// <summary>
    /// >> CodeChangeKind
    /// </summary>
    public enum CodeChangeKind
    {
        
        /// <summary>
        /// >> Active
        /// </summary>
        Active = 0,
        
        /// <summary>
        /// >> Inactive
        /// </summary>
        Inactive = 1,
        
        /// <summary>
        /// >> Reinstrumented
        /// </summary>
        Reinstrumented = 2,
    }
    
    /// <summary>
    /// >> 339 - Variant[gear_common.@event.CodeChangeKind]
    /// </summary>
    public sealed class EnumCodeChangeKind : BaseEnumRust<CodeChangeKind>
    {
        
        /// <summary>
        /// Initializes a new instance of the class.
        /// </summary>
        public EnumCodeChangeKind()
        {
				AddTypeDecoder<Substrate.NetApi.Model.Types.Base.BaseOpt<Substrate.NetApi.Model.Types.Primitive.U32>>(CodeChangeKind.Active);
				AddTypeDecoder<BaseVoid>(CodeChangeKind.Inactive);
				AddTypeDecoder<BaseVoid>(CodeChangeKind.Reinstrumented);
        }
    }
}
