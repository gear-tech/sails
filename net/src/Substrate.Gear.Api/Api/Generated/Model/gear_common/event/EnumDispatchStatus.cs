#nullable disable
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using Substrate.NetApi.Model.Types.Base;
using System.Collections.Generic;


namespace Substrate.Gear.Api.Generated.Model.gear_common.@event
{
    
    
    /// <summary>
    /// >> DispatchStatus
    /// </summary>
    public enum DispatchStatus
    {
        
        /// <summary>
        /// >> Success
        /// </summary>
        Success = 0,
        
        /// <summary>
        /// >> Failed
        /// </summary>
        Failed = 1,
        
        /// <summary>
        /// >> NotExecuted
        /// </summary>
        NotExecuted = 2,
    }
    
    /// <summary>
    /// >> 327 - Variant[gear_common.@event.DispatchStatus]
    /// </summary>
    public sealed class EnumDispatchStatus : BaseEnum<DispatchStatus>
    {
    }
}
