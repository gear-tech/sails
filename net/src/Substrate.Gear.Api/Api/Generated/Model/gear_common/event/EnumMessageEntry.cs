#nullable disable
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using Substrate.NetApi.Model.Types.Base;
using System.Collections.Generic;


namespace Substrate.Gear.Api.Generated.Model.gear_common.@event
{
    
    
    /// <summary>
    /// >> MessageEntry
    /// </summary>
    public enum MessageEntry
    {
        
        /// <summary>
        /// >> Init
        /// </summary>
        Init = 0,
        
        /// <summary>
        /// >> Handle
        /// </summary>
        Handle = 1,
        
        /// <summary>
        /// >> Reply
        /// </summary>
        Reply = 2,
        
        /// <summary>
        /// >> Signal
        /// </summary>
        Signal = 3,
    }
    
    /// <summary>
    /// >> 312 - Variant[gear_common.@event.MessageEntry]
    /// </summary>
    public sealed class EnumMessageEntry : BaseEnumRust<MessageEntry>
    {
        
        /// <summary>
        /// Initializes a new instance of the class.
        /// </summary>
        public EnumMessageEntry()
        {
				AddTypeDecoder<BaseVoid>(MessageEntry.Init);
				AddTypeDecoder<BaseVoid>(MessageEntry.Handle);
				AddTypeDecoder<Substrate.Gear.Api.Generated.Model.gprimitives.MessageId>(MessageEntry.Reply);
				AddTypeDecoder<BaseVoid>(MessageEntry.Signal);
        }
    }
}
