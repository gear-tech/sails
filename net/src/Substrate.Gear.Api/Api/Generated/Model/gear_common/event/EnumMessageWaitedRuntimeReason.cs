#nullable disable
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using Substrate.NetApi.Model.Types.Base;
using System.Collections.Generic;


namespace Substrate.Gear.Api.Generated.Model.gear_common.@event
{
    
    
    /// <summary>
    /// >> MessageWaitedRuntimeReason
    /// </summary>
    public enum MessageWaitedRuntimeReason
    {
        
        /// <summary>
        /// >> WaitCalled
        /// </summary>
        WaitCalled = 0,
        
        /// <summary>
        /// >> WaitForCalled
        /// </summary>
        WaitForCalled = 1,
        
        /// <summary>
        /// >> WaitUpToCalled
        /// </summary>
        WaitUpToCalled = 2,
        
        /// <summary>
        /// >> WaitUpToCalledFull
        /// </summary>
        WaitUpToCalledFull = 3,
    }
    
    /// <summary>
    /// >> 334 - Variant[gear_common.@event.MessageWaitedRuntimeReason]
    /// </summary>
    public sealed class EnumMessageWaitedRuntimeReason : BaseEnum<MessageWaitedRuntimeReason>
    {
    }
}
