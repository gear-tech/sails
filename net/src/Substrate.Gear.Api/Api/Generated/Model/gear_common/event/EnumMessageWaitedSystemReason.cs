#nullable disable
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using Substrate.NetApi.Model.Types.Base;
using System.Collections.Generic;


namespace Substrate.Gear.Api.Generated.Model.gear_common.@event
{
    
    
    /// <summary>
    /// >> MessageWaitedSystemReason
    /// </summary>
    public enum MessageWaitedSystemReason
    {
    }
    
    /// <summary>
    /// >> 335 - Variant[gear_common.@event.MessageWaitedSystemReason]
    /// </summary>
    public sealed class EnumMessageWaitedSystemReason : BaseEnum<MessageWaitedSystemReason>
    {
    }
}
