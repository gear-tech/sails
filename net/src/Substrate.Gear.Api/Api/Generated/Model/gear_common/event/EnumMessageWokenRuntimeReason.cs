#nullable disable
//------------------------------------------------------------------------------
// <auto-generated>
//     This code was generated by a tool.
//
//     Changes to this file may cause incorrect behavior and will be lost if
//     the code is regenerated.
// </auto-generated>
//------------------------------------------------------------------------------

using Substrate.NetApi.Model.Types.Base;
using System.Collections.Generic;


namespace Substrate.Gear.Api.Generated.Model.gear_common.@event
{
    
    
    /// <summary>
    /// >> MessageWokenRuntimeReason
    /// </summary>
    public enum MessageWokenRuntimeReason
    {
        
        /// <summary>
        /// >> WakeCalled
        /// </summary>
        WakeCalled = 0,
    }
    
    /// <summary>
    /// >> 337 - Variant[gear_common.@event.MessageWokenRuntimeReason]
    /// </summary>
    public sealed class EnumMessageWokenRuntimeReason : BaseEnum<MessageWokenRuntimeReason>
    {
    }
}
