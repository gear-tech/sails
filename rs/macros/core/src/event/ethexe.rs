use proc_macro_error::abort;
use proc_macro2::{Span, TokenStream};
use quote::quote;
use syn::{Fields, Ident, ItemEnum, Path, Type, Variant};

pub(super) fn generate_eth_event_impl(input: &ItemEnum, sails_path: &Path) -> TokenStream {
    let enum_ident = &input.ident;

    // Vectors to collect match arms for topics and data.
    let mut sigs_const = Vec::new();
    let mut topics_match_arms = Vec::new();
    let mut data_match_arms = Vec::new();

    // Process each variant.
    for (idx, variant) in input.variants.iter().enumerate() {
        let variant_ident = &variant.ident;
        check_forbidden_event_idents(variant_ident);

        // Prepare pattern, and vectors for indexed and non-indexed field expressions.
        let (pattern, indexed_exprs, non_indexed_exprs): (
            TokenStream,
            Vec<TokenStream>,
            Vec<TokenStream>,
        ) = match &variant.fields {
            // For named fields, use the field identifiers directly.
            Fields::Named(named) => {
                let field_idents: Vec<&Ident> = named
                    .named
                    .iter()
                    .map(|f| f.ident.as_ref().unwrap())
                    .collect();

                // Build the pattern: Enum::Variant { field1, field2, ... }
                let pat = quote! {
                    #enum_ident::#variant_ident { #( #field_idents ),* }
                };

                let mut idx_exprs = Vec::new();
                let mut non_idx_exprs = Vec::new();
                for field in named.named.iter() {
                    let field_ident = field.ident.as_ref().unwrap();
                    if is_indexed(field) {
                        idx_exprs.push(quote! {
                            Self::topic_hash(#field_ident)
                        });
                    } else {
                        non_idx_exprs.push(quote!(#field_ident));
                    }
                }
                if idx_exprs.len() > 3 {
                    abort!(
                        variant,
                        "too many indexed fields (max 3): {}",
                        idx_exprs.len()
                    );
                }
                (pat, idx_exprs, non_idx_exprs)
            }
            // For unnamed (tuple) fields, create synthetic identifiers.
            Fields::Unnamed(unnamed) => {
                if unnamed.unnamed.iter().any(is_indexed) {
                    abort!(
                        unnamed,
                        "unnamed fields cannot be `#[indexed]`, use named fields instead"
                    );
                }

                let non_idx_exprs: Vec<TokenStream> = unnamed
                    .unnamed
                    .iter()
                    .enumerate()
                    .map(|(i, _)| Ident::new(&format!("f{i}"), Span::call_site()))
                    .map(|field_ident| quote!(#field_ident))
                    .collect();
                // Build the pattern: Enum::Variant(f0, f1, ...)
                let pat = quote! {
                    #enum_ident::#variant_ident( #( #non_idx_exprs ),* )
                };

                (pat, vec![], non_idx_exprs)
            }
            // For unit variants, no fields exist.
            Fields::Unit => {
                let pat = quote! {
                    #enum_ident::#variant_ident
                };
                (pat, Vec::new(), Vec::new())
            }
        };

        // Build the signature constant.
        let sig = variant_signature(variant, sails_path);
        sigs_const.push(sig);

        // Build the topics match arm.
        let cap = 1 + indexed_exprs.len();
        let topics_arm = quote! {
            #pattern => {
                let mut topics = #sails_path::Vec::with_capacity( #cap );
                let (_, _, hash) = Self::SIGNATURES[ # idx ];
                topics.push(#sails_path::alloy_primitives::B256::new(hash));
                #( topics.push(#indexed_exprs); )*
                topics
            }
        };
        topics_match_arms.push(topics_arm);

        // Build the data match arm: non-indexed fields are ABI-encoded as a tuple.
        data_match_arms.push(quote! {
            #pattern => {
                Self::encode_sequence(&( #( #non_indexed_exprs, )* ))
            }
        });
    }

    // Generate the implementation for the EthEvent trait.
    quote! {
        impl #sails_path::EthEvent for #enum_ident {
            const SIGNATURES: &'static [#sails_path::gstd::EthEventExpo] = &[
                #( #sigs_const ),*
            ];

            #[allow(unused_variables)]
            fn topics(&self) -> #sails_path::Vec<#sails_path::alloy_primitives::B256> {
                match self {
                    #( #topics_match_arms ),*
                }
            }

            #[allow(unused_variables)]
            fn data(&self) -> #sails_path::Vec<u8> {
                match self {
                    #( #data_match_arms ),*
                }
            }
        }
    }
}

fn is_indexed(field: &syn::Field) -> bool {
    field
        .attrs
        .iter()
        .any(|attr| attr.path().is_ident("indexed"))
}

fn variant_signature(variant: &Variant, sails_path: &Path) -> TokenStream {
    let variant_ident = &variant.ident.to_string();
    let field_types: Vec<&Type> = match &variant.fields {
        Fields::Named(named) => named.named.iter().map(|f| &f.ty).collect(),
        Fields::Unnamed(unnamed) => unnamed.unnamed.iter().map(|f| &f.ty).collect(),
        Fields::Unit => Vec::new(),
    };
    let sol_types = quote! {
        <<( #( #field_types, )* ) as #sails_path::alloy_sol_types::SolValue>::SolType as #sails_path::alloy_sol_types::SolType>::SOL_NAME
    };

    quote! {
        (
            #variant_ident,
            #sol_types,
            #sails_path::keccak_const::Keccak256::new().update(#variant_ident .as_bytes()).update(#sol_types  .as_bytes()).finalize()
        )
    }
}

fn check_forbidden_event_idents(ident: &Ident) {
    const FORBIDDEN_IDENTS: &[&str] = &[
        "ExecutableBalanceTopUpRequested",
        "Message",
        "MessageQueueingRequested",
        "Reply",
        "ReplyQueueingRequested",
        "StateChanged",
        "ValueClaimed",
        "ValueClaimingRequested",
    ];
    if FORBIDDEN_IDENTS.contains(&ident.to_string().as_str()) {
        abort!(
            ident,
            "`{}` is not allowed as an identifier of EthEvent variant, not allowed identifiers are: {:?}",
            ident,
            FORBIDDEN_IDENTS
        );
    }
}

pub(super) fn process_indexed(input: &mut ItemEnum) {
    // Process each variant.
    for variant in input.variants.iter_mut() {
        match &mut variant.fields {
            // For named fields, use the field identifiers directly.
            Fields::Named(named) => {
                for field in named.named.iter_mut().filter(|f| is_indexed(f)) {
                    remove_indexed_and_add_comment(field);
                }
            }
            // For unnamed (tuple) fields, create synthetic identifiers.
            Fields::Unnamed(unnamed) => {
                for field in unnamed.unnamed.iter_mut().filter(|f| is_indexed(f)) {
                    remove_indexed_and_add_comment(field);
                }
            }
            // For unit variants, no fields exist.
            Fields::Unit => {}
        }
    }
}

/// remove indexed attribute from field and add comment
fn remove_indexed_and_add_comment(field: &mut syn::Field) {
    field.attrs.retain(|attr| !attr.path().is_ident("indexed"));
    field.attrs.push(syn::parse_quote! {
        #[doc = r" #[indexed]"]
    });
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn event_process_indexed() {
        // arrange
        let input = quote! {
            pub enum SomeService {
                SomeEvent
                {
                    /// Some comment
                    #[indexed]
                    sender: u128
                },
            }
        };
        let mut input: ItemEnum = syn::parse2(input).unwrap();

        let expected = quote! {
            pub enum SomeService {
                SomeEvent
                {
                    /// Some comment
                    /// #[indexed]
                    sender: u128
                },
            }
        };
        let expected: ItemEnum = syn::parse2(expected).unwrap();

        // act
        process_indexed(&mut input);

        // arrange
        assert_eq!(expected, input);
    }
}
