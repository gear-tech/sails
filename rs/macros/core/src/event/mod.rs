use crate::sails_paths::sails_path_or_default;
use args::{CratePathAttr, SAILS_PATH};
use parity_scale_codec::Encode;
use proc_macro_error::abort;
use proc_macro2::TokenStream;
use quote::quote;
use syn::{Fields, ItemEnum, Path, parse::Parse};

mod args;
#[cfg(feature = "ethexe")]
mod ethexe;

pub fn event(attrs: TokenStream, input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree.
    #[cfg_attr(not(feature = "ethexe"), allow(unused_mut))]
    let mut input: ItemEnum = syn::parse2(input).unwrap_or_else(|err| {
        abort!(
            err.span(),
            "`event` attribute can be applied to enums only: {}",
            err
        )
    });

    // Parse the attributes into a syntax tree.
    let sails_path_attr = syn::parse2::<CratePathAttr>(attrs).ok();
    let sails_path = &sails_path_or_default(sails_path_attr.map(|attr| attr.path()));

    let event_impl = generate_sails_event_impl(&input, sails_path);

    #[cfg(feature = "ethexe")]
    let eth_event_impl = ethexe::generate_eth_event_impl(&input, sails_path);
    #[cfg(feature = "ethexe")]
    ethexe::process_indexed(&mut input);
    #[cfg(not(feature = "ethexe"))]
    let eth_event_impl = quote!();

    quote! {
        #input

        #event_impl

        #eth_event_impl
    }
}

fn generate_sails_event_impl(input: &ItemEnum, sails_path: &Path) -> TokenStream {
    // Parse the input enum
    let enum_ident = &input.ident;
    let variants = &input.variants;
    // Check that the enum has at most 256 variants
    if variants.len() > 256 {
        abort!(
            input,
            "`event` enum can have at most 256 variants, but found {}",
            variants.len()
        )
    }

    // Build match arms for each variant
    let mut match_arms = Vec::new();

    for variant in variants {
        let variant_ident = &variant.ident;
        // Determine the pattern to match this variant, ignoring its fields:
        let pattern = match &variant.fields {
            Fields::Unit => {
                // Unit variant: `Enum::Variant`
                quote! { #enum_ident::#variant_ident }
            }
            Fields::Unnamed(_) => {
                // Tuple variant: `Enum::Variant(..)`
                quote! { #enum_ident::#variant_ident ( .. ) }
            }
            Fields::Named(_) => {
                // Struct variant: `Enum::Variant { .. }`
                quote! { #enum_ident::#variant_ident { .. } }
            }
        };
        // Encode the variant identifier as a sequence of u8
        let encoded_name = variant_ident.to_string().encode();

        // Build the match arm: pattern => &[ ... ],
        let arm = quote! {
            #pattern => &[ #( #encoded_name ),* ]
        };
        match_arms.push(arm);
    }

    // Generate the impl block for `Event`
    quote! {
         impl #sails_path::SailsEvent for #enum_ident {
            fn encoded_event_name(&self) -> &'static [u8] {
                match self {
                    #( #match_arms ),*
                }
            }

            fn skip_bytes() -> usize {
                1 // The first byte is reserved for the index of the event enum variant
            }
        }
    }
}

pub fn derive_sails_event(input: TokenStream) -> TokenStream {
    // Parse the input tokens into a syntax tree.
    let input: ItemEnum = syn::parse2(input).unwrap_or_else(|err| {
        abort!(
            err.span(),
            "`SailsEvent` can only be derived for enums: {}",
            err
        )
    });

    let sails_path_attr = input
        .attrs
        .iter()
        .find(|attr| attr.path().is_ident(SAILS_PATH))
        .map(|attr| {
            attr.parse_args_with(CratePathAttr::parse)
                .unwrap_or_else(|_| abort!(attr, "unexpected value for `crate` argument",))
        });
    let sails_path = &sails_path_or_default(sails_path_attr.map(|attr| attr.path()));

    generate_sails_event_impl(&input, sails_path)
}
