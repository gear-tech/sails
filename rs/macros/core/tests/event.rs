use quote::quote;
use sails_macros_core::event;

#[test]
fn event_basic() {
    let input = quote! {
        pub enum MyEvent {
            MyEvent1,
        }
    };
    let result = event(quote!(), input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn event_sails_rename() {
    let attrs = quote! {
        crate = sails_rename
    };
    let input = quote! {
        pub enum Events {
            MyEvent1 {
                sender: u128,
                amount: u128,
                note: String,
            },
            MyEvent2(u128, u128, String),
            MyEvent3,
        }
    };
    let result = event(attrs, input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}
