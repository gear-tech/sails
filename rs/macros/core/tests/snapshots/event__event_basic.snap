---
source: rs/macros/core/tests/event.rs
expression: result
---
#[derive(sails_rs::Encode, sails_rs::TypeInfo)]
#[codec(crate = sails_rs::scale_codec)]
#[scale_info(crate = sails_rs::scale_info)]
pub enum MyEvent {
    MyEvent1,
}
impl sails_rs::SailsEvent for MyEvent {
    fn encoded_event_name(&self) -> &'static [u8] {
        match self {
            MyEvent::MyEvent1 => {
                &[32u8, 77u8, 121u8, 69u8, 118u8, 101u8, 110u8, 116u8, 49u8]
            }
        }
    }
    fn skip_bytes() -> usize {
        1
    }
}
